### ISSUE TITLE
Check if Data Type implementes Stringable interface for Primary keys

### ISSUE THREAD

**Issue by michnovka** [doctrine/doctrine2#9575]
### Feature Request

I came across [this issue](https://stackoverflow.com/questions/15080573/doctrine-2-orm-datetime-field-in-identifier/71401564#71401564) recently

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | no

#### Summary

I think that if field has defined type in PHP Entity, then Doctrine should check whether this type implements Stringable interface (has __toString() method). If not, an error should be thrown, as UnitOfWork cannot keep track of changes and whole ORM is thus broken.

**Comment by derrabus** [comment]
I wonder if we should rather leverage the type system here. After all, it's the database representation of the ID and the type system could get us that representation.

**Comment by michnovka** [comment]
@derrabus yes, I agree. We could add `Type::convertToPHPString(): string` method or `convertToPHPStringable(): \Stringable`

DBAL would then check if `#[ORM\Id]` is set on column whether this method is implemented and if not, it would throw an error.

A default `Type` implementation could be:
```php
    public function convertToPHPStringable($value, AbstractPlatform $platform): \Stringable
    {
        $value = self::convertToPHPValue($value, $platform);
        
        if(!($value instanceof \Stringable))
            throw new \LogicException("PHP value must implement Stringable interface");
        
        return $value;
    }
```
This would allow for already `Stringable` types to work without any further modification or method overloading.

**Comment by michnovka** [comment]
Since the above would work only for PHP8, we can do instead

```php

    public function convertToPHPString($value, AbstractPlatform $platform): string
    {
        $value = self::convertToPHPValue($value, $platform);

        if (is_object($value) && false === method_exists($value, '__toString')) {
            throw new \LogicException("PHP value must be convertible to string");
        }

        return (string) $value;
    }
```

**Comment by michnovka** [comment]
@derrabus isnt this more for dbal than orm? as we need to edit `Doctrine\DBAL\Types\Type`

**Comment by derrabus** [comment]
Well, the main problem is that naive string cast inside `UnitOfWork`. That's an ORM problem. And no, I don't think we have to change the `Type` class. Its `convertToDatabaseValue()` method should be pretty much what we need.

**Comment by michnovka** [comment]
@derrabus I ran into similar issue again with https://github.com/doctrine/orm/pull/9624

so given your suggestion, why not use the convertToDatabaseValue() function to flatten IDs?

**Comment by beberlei** [comment]
There are some historical reasons why this was not done before, bugs of this kind with DateTime instances have been discussed on this tracker many times.

**Comment by beberlei** [comment]
Onw reasone we dont convert all fields via types to db representation before computing key representation is performance. As this operation is needed many times in each UoW <=> entity interaction .

**Comment by michnovka** [comment]
@beberlei so is the issue with IdentifierFlattener.php ? The performance point you make is true, but at the same time we already cast all value types to string. I do not believe there would be a huge overhead.

At this point it is impossible to e.g. use BackedEnums as IDs. There is no __toString() allowed in BackedEnums. For other custom types, this method can be overloaded and it will work.

**Comment by michnovka** [comment]
After working a bit more on similar issue, I still think that the proper way is to extend DBAL

```php
    public function convertToPHPString($value, AbstractPlatform $platform): string
    {
        $value = self::convertToPHPValue($value, $platform);

        if (is_object($value) && false === method_exists($value, '__toString')) {
            throw new \LogicException("PHP value must be convertible to string");
        }

        return (string) $value;
    }
```

Then let IdentifierFlattener check whether the ID value is stringable. If (and only if) not, fetch the underlying type and call `convertToPHPString()` method.


> Onw reasone we dont convert all fields via types to db representation before computing key representation is performance. As this operation is needed many times in each UoW <=> entity interaction .

This would eliminate this overhead, as it would be called only where necessary (and where currently it would not work and throw exception anyways). Thus the additional work would be done only for case which is currently unsupported. WDYT @beberlei ?
